src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs

```diff
@@ -84,5 +85,51 @@ public static partial class JsonSerializer
 
             return writtenMetadataName;
         }
+
+        internal static bool IsPreserveReferencesEnabled(JsonSerializerOptions options)
+            => options.ReferenceHandler?.UsePreserveSemantics ?? false;
+
+        internal static bool IsIgnoreCyclesEnabled(JsonSerializerOptions options)
+            => !options.ReferenceHandler?.UsePreserveSemantics ?? false;
+
+        internal static bool VerifyObjectCycle(ref WriteStack state, object? obj, JsonSerializerOptions options)
+        {
+            return obj != null && IsIgnoreCyclesEnabled(options) && VerifyObjectCycleNoValidation(ref state, obj);
+        }
+
+        private static bool VerifyObjectCycleNoValidation(ref WriteStack state, object obj)
+        {
+            Stack<ReferenceEqualsWrapper>? stack = state.StackForCycleDetection;
```

Comment:
You could use [`ObjectIdGenerator`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.objectidgenerator?view=net-5.0) instead of creating manual wrappers.

Reply:
I see that the `PreserveReferenceHandler` does something similar by using a dictionary with reference equality comparison. Are there any performance concerns that led you to the current approach?

Reply:
The reason for the `Stack` is that we need to pop the references when we finish writing since the same object can be found in a sibling branch and that doesn't imply circularity.

How could ObjectIDGenerator deal with the case where you have the same reference twice but they are in different branches and no reference loop is involved? e.g:

```cs
class TreeNode 
{
    public string Name { get; set; }
    public TreeNode Left { get; set; }
    public TreeNode Right { get; set; }
}

void Test()
{
    var root = new TreeNode { Name = "Root" };
    var child = new TreeNode { Name = "Child" };

    root.Left = root.Right = child;

    Console.WriteLine(JsonSerializer.Serialize(root, 
        new JsonSerializerOptions { ReferenceHandler = ReferenceHandler.IgnoreCycle }));
    // Prints: {"Name":"Root","Left":{"Name":"Child"},"Right":{"Name":"Child"}}
}
```
It seems to me that the ObjectIDGenerator cannot remove objects that have already been added to it.

Reply:
`ObjectIdGenerator` is merely a class used for generating unique integer id's for each object instance in a graph. You would still need to use a stack to detect cycles, however that stack would simply hold those integers rather than structs wrapping the original objects using reference equality. I would suspect this might result in modest performance gains, though obviously that would need to be measured to be certain.

Reply:
@eiriktsarpalis PTAL at the following.

runtime2's CoreRun uses an `ObjectIDGenerator` and a `Stack<long>` while runtime's CureRun uses the `Stack<ReferenceEqualsWrapper>`, I am using the available classes already used to measure other JSON perf scenarios in MicroBenchmarks. 

You can see that runtime2 takes longer and uses more memory, this might be because `ObjectIDGenerator` will do one extra allocation for each call to JsonSerializer.Serialize. Plus, there's probably more overhead on `Stack<long>.Push(ObjectIDGenerator.GetId(object, out bool))` calls compared to `Stack<ReferenceEqualsWrapper>.Push(new ReferenceEqualsWrapper(object))`.

Here's the runtime branch that implements your suggestion and the benchmarks in case you would like to see if I am missing something.

https://github.com/Jozkee/runtime/commit/29b7825ede8f25cf352220a4744e91a85cf7b2b7
https://github.com/Jozkee/performance/commit/806567df150480dd8c0d253ee072db502acc2263

Also note that in order to use `ObjectIDGenerator`, System.Text.Json needs to take a dependency on System.Runtime.Serialization.Formatters.

``` ini

BenchmarkDotNet=v0.12.1.1466-nightly, OS=Windows 10.0.18363.1256 (1909/November2019Update/19H2)
Intel Core i7-9750H CPU 2.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=5.0.100
  [Host]     : .NET 5.0.0 (5.0.20.51904), X64 RyuJIT
  Job-OZWOXV : .NET 6.0.0 (42.42.42.42424), X64 RyuJIT
  Job-YWJRIT : .NET 6.0.0 (42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:DebugType=portable  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

```
|                                         Type |                           Method |        Job |                                                                                                  Toolchain |         Mean |       Error |      StdDev |       Median |          Min |          Max | Ratio | RatioSD |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|--------------------------------------------- |--------------------------------- |----------- |----------------------------------------------------------------------------------------------------------- |-------------:|------------:|------------:|-------------:|-------------:|-------------:|------:|--------:|--------:|--------:|--------:|----------:|
|             WriteIgnoreCycle&lt;IndexViewModel&gt; | SerializeWithIgnoreCycle_Generic | Job-OZWOXV | \runtime2\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |  24,176.4 ns |   935.79 ns | 1,077.65 ns |  23,860.2 ns |  23,115.9 ns |  26,451.1 ns |  1.16 |    0.06 |  4.1667 |  0.2778 |       - |  26,464 B |
|             WriteIgnoreCycle&lt;IndexViewModel&gt; | SerializeWithIgnoreCycle_Generic | Job-YWJRIT |  \runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |  21,013.8 ns |   181.97 ns |   170.22 ns |  21,025.3 ns |  20,745.7 ns |  21,268.3 ns |  1.00 |    0.00 |  4.0323 |  0.3360 |       - |  25,688 B |
|                                              |                                  |            |                                                                                                            |              |             |             |              |              |              |       |         |         |         |         |           |
|                   WriteIgnoreCycle&lt;Location&gt; | SerializeWithIgnoreCycle_Generic | Job-OZWOXV | \runtime2\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |   1,304.9 ns |    19.67 ns |    17.44 ns |   1,300.9 ns |   1,280.9 ns |   1,339.7 ns |  1.30 |    0.02 |  0.2285 |       - |       - |   1,448 B |
|                   WriteIgnoreCycle&lt;Location&gt; | SerializeWithIgnoreCycle_Generic | Job-YWJRIT |  \runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |   1,003.2 ns |    10.60 ns |     8.85 ns |   1,001.9 ns |     991.6 ns |   1,016.0 ns |  1.00 |    0.00 |  0.1035 |       - |       - |     672 B |
|                                              |                                  |            |                                                                                                            |              |             |             |              |              |              |       |         |         |         |         |           |
|             WriteIgnoreCycle&lt;LoginViewModel&gt; | SerializeWithIgnoreCycle_Generic | Job-OZWOXV | \runtime2\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |     508.6 ns |     7.91 ns |     7.01 ns |     507.3 ns |     501.3 ns |     523.8 ns |  1.16 |    0.03 |  0.1129 |       - |       - |     712 B |
|             WriteIgnoreCycle&lt;LoginViewModel&gt; | SerializeWithIgnoreCycle_Generic | Job-YWJRIT |  \runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |     439.9 ns |     7.22 ns |     6.76 ns |     437.9 ns |     430.9 ns |     453.9 ns |  1.00 |    0.00 |  0.0673 |       - |       - |     432 B |
|                                              |                                  |            |                                                                                                            |              |             |             |              |              |              |       |         |         |         |         |           |
|    WriteIgnoreCycle&lt;MyEventsListerViewModel&gt; | SerializeWithIgnoreCycle_Generic | Job-OZWOXV | \runtime2\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe | 579,134.7 ns | 8,690.61 ns | 7,257.05 ns | 577,576.6 ns | 568,373.4 ns | 592,184.4 ns |  1.10 |    0.01 | 46.8750 | 46.8750 | 46.8750 | 308,982 B |
|    WriteIgnoreCycle&lt;MyEventsListerViewModel&gt; | SerializeWithIgnoreCycle_Generic | Job-YWJRIT |  \runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe | 527,267.9 ns | 4,324.28 ns | 3,610.97 ns | 526,483.8 ns | 522,870.2 ns | 533,368.8 ns |  1.00 |    0.00 | 45.8333 | 45.8333 | 45.8333 | 286,334 B |
|                                              |                                  |            |                                                                                                            |              |             |             |              |              |              |       |         |         |         |         |           |
|            WriteIgnoreCycle&lt;SimpleListOfInt&gt; | SerializeWithIgnoreCycle_Generic | Job-OZWOXV | \runtime2\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |     310.6 ns |     3.58 ns |     3.17 ns |     309.4 ns |     307.0 ns |     316.9 ns |  1.14 |    0.02 |  0.0899 |       - |       - |     568 B |
|            WriteIgnoreCycle&lt;SimpleListOfInt&gt; | SerializeWithIgnoreCycle_Generic | Job-YWJRIT |  \runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |     273.6 ns |     2.00 ns |     1.67 ns |     273.5 ns |     271.3 ns |     277.2 ns |  1.00 |    0.00 |  0.0453 |       - |       - |     288 B |
|                                              |                                  |            |                                                                                                            |              |             |             |              |              |              |       |         |         |         |         |           |
| WriteIgnoreCycle&lt;SimpleStructWithProperties&gt; | SerializeWithIgnoreCycle_Generic | Job-OZWOXV | \runtime2\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |     327.4 ns |     2.39 ns |     2.24 ns |     327.6 ns |     324.3 ns |     330.5 ns |  1.14 |    0.02 |  0.0894 |       - |       - |     568 B |
| WriteIgnoreCycle&lt;SimpleStructWithProperties&gt; | SerializeWithIgnoreCycle_Generic | Job-YWJRIT |  \runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe |     287.0 ns |     3.86 ns |     3.61 ns |     285.6 ns |     283.4 ns |     293.5 ns |  1.00 |    0.00 |  0.0450 |       - |       - |     288 B |

Reply:
Got it, thanks for taking the time to investigate.